trustpinning/certs_test.go

```diff
@@ -781,6 +781,200 @@ func testValidateRootRotationMissingNewSig(t *testing.T, keyAlg, rootKeyType str
+	require.Error(t, err)
+
+	// This call will succeed since we include the new root cert ID (and the old one)
+	validCertConfig := trustpinning.TrustPinConfig{
+		Certs: map[string][]string{
+			gun: {origRootKey.ID(), replRootKey.ID()},
+		},
+		DisableTOFU: true,
+	}
+	validatedRoot, err := trustpinning.ValidateRoot(prevRoot, signedTestRoot, gun, validCertConfig)
+	require.NoError(t, err)
+	generateRootKeyIDs(typedSignedRoot)
+	require.Equal(t, typedSignedRoot, validatedRoot)
+
+	// This call will also succeed since we only need the new replacement root ID to be pinned
+	validCertConfig = trustpinning.TrustPinConfig{
+		Certs: map[string][]string{
+			gun: {replRootKey.ID()},
+		},
+		DisableTOFU: true,
+	}
+	validatedRoot, err = trustpinning.ValidateRoot(prevRoot, signedTestRoot, gun, validCertConfig)
+	require.NoError(t, err)
+	generateRootKeyIDs(typedSignedRoot)
+	require.Equal(t, typedSignedRoot, validatedRoot)
+
+	// Even if we disable TOFU in the trustpinning, since we have a previously trusted root we should honor a valid rotation
+	validatedRoot, err = trustpinning.ValidateRoot(prevRoot, signedTestRoot, gun, trustpinning.TrustPinConfig{DisableTOFU: true})
+	require.NoError(t, err)
+	generateRootKeyIDs(typedSignedRoot)
+	require.Equal(t, typedSignedRoot, validatedRoot)
+}
+
+// TestValidateRootRotationTrustPinningInvalidCA runs a full root certificate rotation but ensures that
+// the specified trust pinning rejectas the new root for not being signed by the specified CA
+func TestValidateRootRotationTrustPinningInvalidCA(t *testing.T) {
+	gun := "notary-signer"
+	keyAlg := data.RSAKey
+	// Temporary directory where test files will be created
+	tempBaseDir, err := ioutil.TempDir("", "notary-test-")
+	defer os.RemoveAll(tempBaseDir)
+	require.NoError(t, err, "failed to create a temporary directory: %s", err)
+
+	leafCert, err := trustmanager.LoadCertFromFile("../fixtures/notary-signer.crt")
+	require.NoError(t, err)
+
+	intermediateCert, err := trustmanager.LoadCertFromFile("../fixtures/intermediate-ca.crt")
+	require.NoError(t, err)
+
+	pemChainBytes, err := trustmanager.CertChainToPEM([]*x509.Certificate{leafCert, intermediateCert})
+	require.NoError(t, err)
+
+	origRootKey := data.NewPublicKey(data.RSAx509Key, pemChainBytes)
+
+	rootRole, err := data.NewRole(data.CanonicalRootRole, 1, []string{origRootKey.ID()}, nil)
+	require.NoError(t, err)
+
+	testRoot, err := data.NewRoot(
+		map[string]data.PublicKey{origRootKey.ID(): origRootKey},
+		map[string]*data.RootRole{
+			data.CanonicalRootRole:      &rootRole.RootRole,
+			data.CanonicalTimestampRole: &rootRole.RootRole,
+			data.CanonicalTargetsRole:   &rootRole.RootRole,
+			data.CanonicalSnapshotRole:  &rootRole.RootRole},
+		false,
+	)
+	testRoot.Signed.Version = 1
+	require.NoError(t, err, "Failed to create new root")
+
+	keyReader, err := os.Open("../fixtures/notary-signer.key")
+	require.NoError(t, err, "could not open key file")
+	pemBytes, err := ioutil.ReadAll(keyReader)
```

Comment:
Non-blocking nit:  The open + `ioutil.ReadAll` could also be replaced with `ioutil.ReadFile`.